File: src/main/java/gtlp/groundmc/customrecipes/listeners/DisenchantmentListener.java
package gtlp.groundmc.customrecipes.listeners;

import org.bukkit.Material;
import org.bukkit.enchantments.Enchantment;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.inventory.InventoryAction;
import org.bukkit.event.inventory.InventoryClickEvent;
import org.bukkit.event.inventory.PrepareAnvilEvent;
import org.bukkit.inventory.AnvilInventory;
import org.bukkit.inventory.ItemStack;
import org.bukkit.inventory.meta.EnchantmentStorageMeta;

import java.util.Map;

/**
 * Listener that handles events directed at anvils and allows enchanted items
 * to be disenchanted.
 */
public class DisenchantmentListener implements Listener {

    /**
     * Prepares the disenchantment of an item.
     *
     * @param event the event to handle
     */
    @EventHandler
    public void prepareDisenchant(PrepareAnvilEvent event) {
        AnvilInventory inventory = event.getInventory();
        ItemStack firstItem = inventory.getItem(0);
        ItemStack book = inventory.getItem(1);
        if (firstItem != null &&
                !firstItem.getEnchantments().isEmpty() &&
                book != null &&
                book.getType() == Material.BOOK_AND_QUILL) {

            Map.Entry<Enchantment, Integer> enchantment =
                    firstItem.getEnchantments().entrySet().iterator().next();

            ItemStack enchantedBook = new ItemStack(Material.ENCHANTED_BOOK);
            EnchantmentStorageMeta meta =
                    (EnchantmentStorageMeta) enchantedBook.getItemMeta();
            meta.addStoredEnchant(
                    enchantment.getKey(),
                    enchantment.getValue(),
                    false
            );
            enchantedBook.setItemMeta(meta);
            event.setResult(enchantedBook);
            inventory.setRepairCost(enchantment.getValue() * 2);
        }
    }

    /**
     * Actually disenchants the item and gives the player with the permission
     * to do so an enchanted book with the enchantment on it.
     * Removes the enchantment from the item in the first slot and converts an
     * ordinary book and quill into an enchanted book.
     *
     * @param event the event to handle
     */
    @EventHandler
    public void disenchant(InventoryClickEvent event) {
        Player player;
        if (event.getWhoClicked() instanceof Player) {
            player = (Player) event.getWhoClicked();
        } else {
            return;
        }
        if (!player.hasPermission("customrecipes.disenchant")) {
            return;
        }
        if (event.getAction() == InventoryAction.NOTHING &&
                event.isLeftClick() &&
                event.getClickedInventory() instanceof AnvilInventory) {
            AnvilInventory inventory =
                    (AnvilInventory) event.getClickedInventory();
            ItemStack firstItem = inventory.getItem(0);
            ItemStack book = inventory.getItem(1);
            if (firstItem != null &&
                    !firstItem.getEnchantments().isEmpty() &&
                    book != null &&
                    book.getType() == Material.BOOK_AND_QUILL) {
                Map.Entry<Enchantment, Integer> enchantment =
                        firstItem.getEnchantments().entrySet().iterator().next();

                if (player.getLevel() < enchantment.getValue() * 2) {
                    return;
                }

                book.setAmount(book.getAmount() - 1);

                firstItem.removeEnchantment(enchantment.getKey());

                inventory.setItem(0, firstItem);
                inventory.setRepairCost(enchantment.getValue() * 2);
                player.setLevel(player.getLevel() - (enchantment.getValue() * 2));

                ItemStack enchantedBook = new ItemStack(Material.ENCHANTED_BOOK);
                EnchantmentStorageMeta meta =
                        (EnchantmentStorageMeta) enchantedBook.getItemMeta();
                meta.addStoredEnchant(
                        enchantment.getKey(),
                        enchantment.getValue(),
                        false
                );
                enchantedBook.setItemMeta(meta);

                event.setCurrentItem(enchantedBook);
                event.getView().setCursor(enchantedBook);
                event.setCancelled(true);
            }
        }
    }
}

File: src/main/kotlin/gtlp/groundmc/customrecipes/listeners/DisenchantmentListener.kt
package gtlp.groundmc.customrecipes.listeners

import org.bukkit.Material
import org.bukkit.entity.Player
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.event.inventory.InventoryAction
import org.bukkit.event.inventory.InventoryClickEvent
import org.bukkit.event.inventory.PrepareAnvilEvent
import org.bukkit.inventory.AnvilInventory
import org.bukkit.inventory.ItemStack
import org.bukkit.inventory.meta.EnchantmentStorageMeta

/**
 * Listener that handles events directed at anvils and allows enchanted items
 * to be disenchanted.
 */
class DisenchantmentListener : Listener {

    /**
     * Prepares the disenchantment of an item.
     *
     * @param event the event to handle
     */
    @EventHandler
    fun prepareDisenchant(event: PrepareAnvilEvent) {
        val inventory = event.inventory
        val firstItem = inventory.getItem(0)
        val book = inventory.getItem(1)
        if (firstItem != null &&
                !firstItem.enchantments.isEmpty() &&
                book != null &&
                book.type == Material.BOOK_AND_QUILL) {

            val enchantment = firstItem.enchantments.entries.iterator().next()

            val enchantedBook = ItemStack(Material.ENCHANTED_BOOK)
            val meta = enchantedBook.itemMeta as EnchantmentStorageMeta
            meta.addStoredEnchant(
                    enchantment.key,
                    enchantment.value,
                    false
            )
            enchantedBook.itemMeta = meta
            event.result = enchantedBook
            inventory.repairCost = enchantment.value * 2
        }
    }

    /**
     * Actually disenchants the item and gives the player with the permission
     * to do so an enchanted book with the enchantment on it.
     * Removes the enchantment from the item in the first slot and converts an
     * ordinary book and quill into an enchanted book.
     *
     * @param event the event to handle
     */
    @EventHandler
    fun disenchant(event: InventoryClickEvent) {
        val player = event.whoClicked as Player
        if (!player.hasPermission("customrecipes.disenchant")) {
            event.isCancelled = true
            return
        }
        if (event.action == InventoryAction.NOTHING &&
                event.isLeftClick &&
                event.clickedInventory is AnvilInventory) {
            val inventory = event.clickedInventory as AnvilInventory
            val firstItem = inventory.getItem(0)
            val book = inventory.getItem(1)
            if (firstItem != null &&
                    !firstItem.enchantments.isEmpty() &&
                    book != null &&
                    book.type == Material.BOOK_AND_QUILL) {
                val enchantment = firstItem.enchantments.entries.iterator().next()

                if (player.level < enchantment.value * 2) {
                    event.isCancelled = true
                    return
                }

                book.amount--

                firstItem.removeEnchantment(enchantment.key)

                inventory.setItem(0, firstItem)
                inventory.repairCost = enchantment.value * 2
                player.level = player.level - enchantment.value * 2

                val enchantedBook = ItemStack(Material.ENCHANTED_BOOK)
                val meta = enchantedBook.itemMeta as EnchantmentStorageMeta
                meta.addStoredEnchant(
                        enchantment.key,
                        enchantment.value,
                        false
                )
                enchantedBook.itemMeta = meta

                event.currentItem = enchantedBook
                event.view.cursor = enchantedBook
                event.isCancelled = true
            }
        }
    }
}
